### ISSUE TITLE
Spy on console.error doesn't work

### ISSUE THREAD

**Issue by ludovicmnji** [testing-library/react-hooks-testing-library#555]
- `react-hooks-testing-library` version: 5.0.3
- `react` version: 16.13.1
- `node` version: 14.15.4

### Relevant code or config:

```js
    it('should abort request if component is unmounting', async () => {
      // Given
      const consoleErrorSpy = jest.spyOn(global.console, 'error');
      const { result, unmount } = renderHook(() => useFetch('key', fetchFn));

      // When
      unmount();

      // Then
      expect(fetchFn).toHaveBeenCalledTimes(1);
      expect(result.current.status).toBe('pending');
      expect(consoleErrorSpy).not.toHaveBeenCalled();
    });
```

### What you did:

I try to make a test failed when I get a React warning. For example, I run a race condition in my source code and I get the following error in console : `Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.`

### What happened:

This test always passed, even if the console.error is shown in the terminal  so it's a false positive test. 

<img width="1399" alt="Capture d’écran 2021-01-25 à 16 40 31" src="https://user-images.githubusercontent.com/6171841/105728168-204c4080-5f2c-11eb-9e57-d77eb216740b.png">

**Comment by joshuaellis** [comment]
Hi @ludovicmnji,

Would you be able to share the code for the hook? It's hard for me to figure out whats going on & create a local env to test in without the hook.

**Comment by ludovicmnji** [comment]
Hi @joshuaellis,

Here a Codesandbox with the portion of code https://codesandbox.io/s/console-error-rtl-b2u76 (the test didn't pass because of strange error)

**Comment by joshuaellis** [comment]
Thanks, someone will look as soon as they can. In the meantime, I think you're spying on `console` wrong, you don't need to write `global.console` I believe you can just write `console`. At least I have in my tests before.

**Comment by ludovicmnji** [comment]
Yes I tried with just `console` but it's the same.

Thanks

**Comment by joshuaellis** [comment]
Hi @ludovicmnji you've changed the sandbox so I can't see anything you want me to look at?

**Comment by ludovicmnji** [comment]
I simplified the hook that produce the issue. I called it `useMemoryLeak`. I've just rewritten the original code if you want it.

**Comment by joshuaellis** [comment]
I don't think sandbox is the best place for testing, it seems like `mock` & `useFakeTimers` don't work.  see here – https://github.com/codesandbox/codesandbox-client/issues/513. So, i've redone the sandbox in a repo here – https://github.com/joshuaellis/rhtl-555 and the test passes with no errors.

Maybe i'm missing something?

**Comment by ludovicmnji** [comment]
@joshuaellis Could you pass your repositories in public ?

**Comment by joshuaellis** [comment]
sorry! I didn't realise I made it private.

**Comment by mpeyper** [comment]
Can I just clarify the issue @ludovicmnji, did this test pass prior to version 5.0.3 with or without the console logs or are you trying to write a new test that you expect to fail, but is passing?

Generally the logs shown there are commonly seen when the test is not waiting for the update to occur, and the assertions pass, the test finishes and the test runner moves on to the next year. While the next test is running, the async behaviour resolves and updates the unmounted component outside of any `act` logic.

**Comment by ludovicmnji** [comment]
@joshuaellis You have to comment [this line](https://github.com/joshuaellis/rhtl-555/blob/main/src/useMemoryLeak.js#L29 ) to make the test passed **with** error logs while the test expect to fail because console.error had been called.

@mpeyper No matter the version of `@testing-library/react-hooks` (I tried with 5.0.0 before), I expect the test failed because it displays the React warning. It's the use case I want to be sure to not have in production if someone refactor the hook someday. Maybe I don't make it in the right way? But to clarify, I want to avoid this sequence :
1. The hook trigger a pending request
2. The request is still pending but the component is unmounting
3. The request is in success and trigger a `setState` / `dispatch` on the unmounted component = **memory leak**

**Comment by mpeyper** [comment]
Try putting a few console logs in your test and your hook. I like to number them, e.g. `console.log(1)`, `console.log(2)`, ... , in the order I expect them to show. You may be surprised when some things are happening.

**Comment by ludovicmnji** [comment]
Thanks for the tip. My hook is working properly.  What I want to achieve it's having a robust test suites without false positive (for CI mainly) and without testing implementation too deeply.

**Comment by joshuaellis** [comment]
The hook isn't going to fail because it shows:
```shell
    Warning: An update to TestComponent inside a test was not wrapped in act(...).
    
    When testing, code that causes React state updates should be wrapped into act(...):
    
    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */
```

If you want to test that your component did not fire a the dispatch call when you unmount __before__ the promise resolves it's self, you should probably look at testing that an async function such as `waitForNextUpdate` fails. Because there should be no update.

Testing if "The request is still pending but the component is unmounting" you need to have something to test again, I don't think theres a way to spy on pending promises in JS, so maybe the hook should be returning a pending state and you can test against that.

Testing if "The hook trigger a pending request", well that's already been done by seeing if `fn()` is being called

In short, I don't think this is a bug....

**Comment by ludovicmnji** [comment]
> If you want to test that your component did not fire a the dispatch call when you unmount before the promise resolves it's self, you should probably look at testing that an async function such as waitForNextUpdate fails. Because there should be no update.

How would you write that concretely in the example ? Just putting `waitForNextUpdate()` at the end of the test will not make the test case failed.

> Testing if "The request is still pending but the component is unmounting" you need to have something to test again, I don't think theres a way to spy on pending promises in JS, so maybe the hook should be returning a pending state and you can test against that.

Yes, I have a pending state and have an expect dedicated for this. I didn't write in the example to be more focus on console.error issue.

**Comment by joshuaellis** [comment]
> Yes, I have a pending state and have an expect dedicated for this. I didn't write in the example to be more focus on console.error issue.

the console.error isnt an issue, so it's really hard to help you write your test if all the bits aren't there to actually write the test for.

There are loads of examples in closed issues how to tackle what you want to do, here's one: https://github.com/testing-library/react-hooks-testing-library/issues/425 this person uses [manual mocks](https://jestjs.io/docs/en/manual-mocks) you could mock react's useReducer function & test that `dispatch` **does not fire**. This would be more reliable.

**Comment by mpeyper** [comment]
@ludovicmnji how about something like this:

```js
import { renderHook } from '@testing-library/react-hooks'
import useMemoryLeak from '../useMemoryLeak'

const fetchFn = jest.fn().mockImplementation(
  () =>
    new Promise((resolve) => {
      setTimeout(() => resolve("fetch"), 200);
    })
);

test('should abort request if component is unmounting', async () => {
  // Given
  const consoleErrorSpy = jest.spyOn(console, 'error')
  const { unmount , waitForNextUpdate } = renderHook(() => useMemoryLeak(fetchFn))

  // When
  unmount()

  // Then
  await expect(waitForNextUpdate({ timeout: 250 })).rejects.toEqual(Error('Timed out in waitForNextUpdate after 250ms.'))

  expect(consoleErrorSpy).not.toHaveBeenCalled()
  expect(fetchFn).toHaveBeenCalled()
})
```

This works by waiting for longer than the `fetchFn` timeout for the component to update before it checks for the console output.  This test passes when the `isCancelled` flag is set in the effect cleanup, and fails on `expect(consoleErrorSpy).not.toHaveBeenCalled()` when it is commented out.

If you've got a status variable to check for in the real test, you could replace `waitForNextUpdate` to use `waitForValueToChange` instead if you want, e.g. 

```js
await expect(waitForValueToChange(() => result.current.status, { timeout: 250 })).rejects.toEqual(Error('Timed out in waitForValueToChange after 250ms.'))
```

I'm not sure if that provides any more value that the previous example though.  I'll leave that up to you to determine.

**Comment by ludovicmnji** [comment]
First, thanks for you help and your time.

@joshuaellis : I added the status pending to be closer to the real hook, but in my opinion it doesn't change so much my problem for writing the tetst. I completed with @mpeyper suggestions also (see after). In addition, I would like to avoid testing implementation because if someone wants to change the implementation, I would prefer he will not need to rewrite the test. In the end, if there is no solution, I will give up.

```js
// useMemoryLeak.js
import { useEffect, useReducer } from "react";

function reducer(state, action) {
  switch (action.type) {
    case "PENDING":
      return { status: "pending", data: null };
    case "SUCCESS":
      return { status: "success", data: action.payload };
    default:
      return state;
  }
}

const initialState = { status: "idle", data: null };

export default function useMemoryLeak(fetchFn) {
  const [state, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    let isCancelled = false;

    async function waitAndDispatch() {
      dispatch({ type: "PENDING" });

      try {
        const data = await fetchFn();
        if (!isCancelled) {
          dispatch({ type: "SUCCESS", payload: data });
        }
      } catch (error) {
        throw error;
      }
    }

    waitAndDispatch();

    return () => {
      /* FIXME: uncomment should make the test passed (green) / comment should make the test failed (red) */
      // isCancelled = true;
    };
  }, [fetchFn]);

  return state;
}
```

```js
// __tests__/useMemoryLeak.js
import { renderHook } from "@testing-library/react-hooks";
import useMemoryLeak from "../useMemoryLeak";

jest.useFakeTimers("modern");

const fetchFn = jest.fn().mockImplementation(
  () =>
    new Promise((resolve) => {
      setTimeout(() => resolve("fetch"), 200);
    })
);

test("should abort request if component is unmounting", async () => {
  // Given
  // const consoleErrorSpy = jest.spyOn(console, 'error')
  const { result, unmount, waitForValueToChange } = renderHook(() =>
    useMemoryLeak(fetchFn)
  );

  // When
  unmount();

  // Then
  await expect(
    waitForValueToChange(() => result.current.status, { timeout: 250 })
  ).rejects.toEqual(Error("Timed out in waitForValueToChange after 250ms."));
  // expect(consoleErrorSpy).not.toHaveBeenCalled()
  expect(fetchFn).toHaveBeenCalled();
  expect(result.current.status).toBe("pending");
});
```

@mpeyper : With `waitForValueToChange` (or no matter, `waitFornextUpdate`), I get the error `Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Timeout` even if I put `isCancelled = true` (that should make the test passed). So the test failed.

**Comment by joshuaellis** [comment]
@ludovicmnji I'm really sorry, I don't get what you want to test in this case, if you want to test if the hook didn't run a dispatch action because it was unmounted & you're opposed to testing if `dispatch` was called, then what are you aiming to test? You can't check for the no-op because that's react implementation and imo unreliable. If you're going to update code, it's pretty reasonable to expect to have to change the tests, not necessarily re-write them, but at the very least update them.

**Comment by ludovicmnji** [comment]
@joshuaellis I don't know how to explain my problem differently. That's why my original request was to watch console.error because it's the most reliable imo (more than testing implementation inside the hook). A portion of code that trigger a console.error (but I understand that is React implementation) during testing phase should not be green.

**Comment by mpeyper** [comment]
@ludovicmnji please note that in my example, I had removed the `jest.useFakeTimers("modern");` line from the test (so the timeout would fire itself).

**Comment by ludovicmnji** [comment]
@mpeyper You're totally right (I updated the example in case someone read that). I don't have the timeout error anymore but the test is still green (with React no-op warning) if the cleanup `isCancelled = true` is commented (that's my point).

**Comment by mpeyper** [comment]
That is not what I'm seeing...

Flag commented out:
![image](https://user-images.githubusercontent.com/23029903/105846414-42c78200-6030-11eb-83f1-395fe8b57c27.png)

Flag left in:
![image](https://user-images.githubusercontent.com/23029903/105846484-5d016000-6030-11eb-87df-f849f4e9161c.png)

This is with my `waitForNextUpdate` version.  replacing the test with your `waitForValueToChange` version does produce the false positive.

**Comment by ludovicmnji** [comment]
My mistake, I was forget the `consoleSpyError` expect. This solution is very acceptable, thank you, that is what I want to achieve. 

Async utils are very difficult to understand. I'm not sure to understand completely the solution anyway but you teach me a new way to handle timers with RTL.

**Comment by mpeyper** [comment]
Asserting something _didn't_ occur is always tough because there is other way (that I'm aware of) than just waiting long enough for that you're confident an update would have happened if it was going to happen.  That's what this solution does.  It sets up an update to occur in 200ms, then it unmounts the component and waits for 250ms, then check that no logs were produced.

If an update happens (can only occur if the component failed to unmount for some reason - I'm not sure if that's even possible), `waitForNextUpdate` wont `reject`, so that assertion will fail.  If and update was attempted on the unmounted component, the log will be produced at around 200ms and `waitForNextUpdate` will reject at around 250ms, passing that assertion but failing the next one checking `console.error` spy.

It might have been possible to get this test working with a fake timer, but In general I don't like faking timers.  I've seen more cases where the tests get harder to read or become fiddly to get passing than I've seen the extra 250ms of test runtime become an issue, plus it gives me more confidence that it's going to work in the real world where time can't be controlled by calling a function.

**Comment by ludovicmnji** [comment]
Ok thanks for explanations it's clearer ✨
